### PR TITLE
fixing the fallback to get the config to access Vault

### LINKED ISSUE #124: folder-scoped approle is not utilized when retrieving folder-scoped credentials
Observed:
When a folder is configured to use a specific approle (different from the globally configured approle) the credentials which are also scoped to the folder (_used for SCM checkout_) are retrieved through the global approle instead of the one which the folder is configured to use. 

To reproduce:
1. Create a new folder in jenkins with approle credential inside this folder along with vault credentials (e.g. vault string credential)
2. Configure folder property to use this new approle (and **not** the global approle)
3. Create new pipeline (with bitbucket/github scm) which uses vault credentials + approle from this folder
4. Observe error  (such as: `FAILED to retrieve username key:
java.lang.RuntimeException: com.datapipe.jenkins.vault.exception.VaultPluginException: Key folder_specific_key could not be found in path secret/folder/credentials`

This seems to prevent the ability to use per-folder approle / credentials for pipeline checkouts.

### LINKED ISSUE #134: FolderVaultConfiguration not working
Yep 

I have a problem configuring the plugin. I don't want a global configuration. In each project (cloudbees-folder) I would like to configure a credential (appRole) to access the Vault.
  It doesn't seem to work

Different tests:
global configuration without url and credential (appRole)
Global configuration with url without credential (appRole)
Global configuration with url and identifier without rights

Capture d’écran 2020-10-09 à 13 04 04
I would like to create the credentials for each project so that they can continue to use credential ID (to switch to vault) and / or substitute the old stored locally : no breakingchange in the pipeline which continues to use IDs
In project folder :
Capture d’écran 2020-10-09 à 13 10 54

Only global configuration with url and credential (admin) works but projects inherit this configuration (which I don't want)
For all the tets, the plugin configuration is done in the project folder with the right url and the right credential
I can't find this behavior when using the plugin in a freestyle job
can you help me ?
Thank you

### LINKED ISSUE #191: Vault Username-Password Credential does not work within Folder level
I set up Jenkins integration with Hashicorp Vault via a plugin. Faced with the fact that when using the "Vault Username-Password Credential" plugin ignores Folder settings and uses only plugin global settings. If the global settings (/configure) of the plugin are not specified, then retrieving the secret through the "Vault Username-Password Credential" is impossible.

### Version report

Jenkins and plugins versions report: Jenkins: 2.289.2 
OS: Linux - 3.10.0-1160.11.1.el7.x86_64

---
ace-editor:1.1
active-directory:2.24
allure-jenkins-plugin:2.29.0
ansicolor:1.0.0
ant:1.11
antisamy-markup-formatter:2.1
apache-httpcomponents-client-4-api:4.5.13-1.0
authentication-tokens:1.4
basic-branch-build-strategies:1.3.2
blueocean:1.24.7
blueocean-autofavorite:1.2.4
blueocean-bitbucket-pipeline:1.24.7
blueocean-commons:1.24.7
blueocean-config:1.24.7
blueocean-core-js:1.24.7
blueocean-dashboard:1.24.7
blueocean-display-url:2.4.1
blueocean-events:1.24.7
blueocean-git-pipeline:1.24.7
blueocean-github-pipeline:1.24.7
blueocean-i18n:1.24.7
blueocean-jira:1.24.7
blueocean-jwt:1.24.7
blueocean-personalization:1.24.7
blueocean-pipeline-api-impl:1.24.7
blueocean-pipeline-editor:1.24.7
blueocean-pipeline-scm-api:1.24.7
blueocean-rest:1.24.7
blueocean-rest-impl:1.24.7
blueocean-web:1.24.7
bootstrap4-api:4.6.0-3
bootstrap5-api:5.0.1-2
bouncycastle-api:2.20
branch-api:2.6.4
build-token-trigger:1.0.0
build-user-vars-plugin:1.7
build-with-parameters:1.5.1
caffeine-api:2.9.1-23.v51c4e2c879c8
checks-api:1.7.0
cloudbees-bitbucket-branch-source:2.9.9
cloudbees-folder:6.15
command-launcher:1.6
conditional-buildstep:1.4.1
config-file-provider:3.8.0
copyartifact:1.46.1
credentials:2.5
credentials-binding:1.26
dashboard-view:2.17
display-url-api:2.3.5
docker-commons:1.17
docker-workflow:1.26
downstream-buildview:1.9
durable-task:1.37
echarts-api:5.1.2-2
email-ext:2.83
external-monitor-job:1.7
favorite:2.3.3
font-awesome-api:5.15.3-3
fortify:20.2.35
generic-webhook-trigger:1.74
git:4.7.2
git-client:3.7.2
git-parameter:0.9.13
git-server:1.9
github:1.33.1
github-api:1.123
github-branch-source:2.11.1
global-build-stats:1.5
handlebars:3.0.8
handy-uri-templates-2-api:2.1.8-1.0
hashicorp-vault-plugin:3.8.0
hp-application-automation-tools-plugin:6.9
htmlpublisher:1.25
influxdb:3.0.1
jackson2-api:2.12.3
javadoc:1.6
jdk-tool:1.5
jenkins-design-language:1.24.7
jira:3.5
jira-steps:1.6.0
jira-trigger:1.0.1
jjwt-api:0.11.2-9.c8b45b8bb173
job-restrictions:0.8
jquery:1.12.4-1
jquery-detached:1.2.1
jquery3-api:3.6.0-1
jsch:0.1.55.2
junit:1.51
kubernetes:1.30.0
kubernetes-client-api:5.4.1
kubernetes-credentials:0.9.0
ldap:2.7
leastload:3.0.0
lockable-resources:2.11
mailer:1.34
mask-passwords:3.0
matrix-auth:2.6.7
matrix-project:1.19
maven-plugin:3.12
mercurial:2.15
metrics:4.0.2.8
momentjs:1.1.1
monitoring:1.87.0
nexus-jenkins-plugin:3.11.20210716-075132.3b66565
nodelabelparameter:1.8.1
okhttp-api:3.14.9
pam-auth:1.6
parameterized-trigger:2.41
pipeline-build-step:2.13
pipeline-graph-analysis:1.11
pipeline-input-step:2.12
pipeline-milestone-step:1.3.2
pipeline-model-api:1.8.5
pipeline-model-declarative-agent:1.1.1
pipeline-model-definition:1.8.5
pipeline-model-extensions:1.8.5
pipeline-multibranch-defaults:2.1
pipeline-rest-api:2.19
pipeline-stage-step:2.5
pipeline-stage-tags-metadata:1.8.5
pipeline-stage-view:2.19
pipeline-utility-steps:2.8.0
plain-credentials:1.7
plugin-util-api:2.3.0
popper-api:1.16.1-2
popper2-api:2.5.4-2
prisma-cloud-jenkins-plugin:20.12.541
prometheus:2.0.10
pubsub-light:1.16
purge-job-history:1.6
rebuild:1.32
repository-connector:2.0.5
resource-disposer:0.16
role-strategy:3.1.1
run-condition:1.5
saml:2.0.7
scm-api:2.6.4
script-security:1.77
snakeyaml-api:1.29.1
sse-gateway:1.24
ssh-credentials:1.19
ssh-slaves:1.32.0
sshd:3.0.3
stashNotifier:1.20
structs:1.23
timestamper:1.13
token-macro:2.15
trilead-api:1.0.13
uno-choice:2.5.6
variant:1.4
windows-slaves:1.8
workflow-aggregator:2.6
workflow-api:2.46
workflow-basic-steps:2.23
workflow-cps:2.92
workflow-cps-global-lib:2.21
workflow-durable-task-step:2.39
workflow-job:2.41
workflow-multibranch:2.26
workflow-scm-step:2.13
workflow-step-api:2.23
workflow-support:3.8

### Reproduction steps

1. Install plugin HashiCorp Vault (https://plugins.jenkins.io/hashicorp-vault-plugin/)
2. Configure plugin in Jenkins in global settings (/configure): set Url and Credentials for auth in Vault (**vault root token**, but i try approle too)
3. Create in Jenkins credentials type "**Vault Username-Password Credential**": set path to secret, username and password key, engine version, credentialsId = **My-HvaultCred** and other. Press "Test Vault Secrets retrieval" button and it reports: "Successfully retrieved username ******** and the password"
4. Test this credentials in pipeline (**MyFolder**/MyPipeline) and it works ok:
`checkout([$class: 'GitSCM', branches: [[name: "master"]], doGenerateSubmoduleConfigurations: false, extensions: [[$class: 'CloneOption', depth: 1, noTags: true, shallow: false]], submoduleCfg: [], userRemoteConfigs: [[credentialsId: "**My-HvaultCred**", url: "${APP_REPO_URL}"]]])`
5. Clean plugin global settings (/configure)
6. Configure Folder settings **MyFolder** where pipeline is: set the same properties that i set in global (same url and same vault root token)
7. I try to check retrieval like in step 3, but it failed.
> FAILED to retrieve username key:
com.datapipe.jenkins.vault.exception.VaultPluginException: The credential id was not configured - please specify the credentials to use.
8. Start pipeline - it failed too.

> [Pipeline] checkout
[2021-09-07T15:24:13.565Z] The recommended git tool is: git
[2021-09-07T15:24:13.573Z] using credential *******-service-hvaultid-****-1
[2021-09-07T15:24:13.580Z] Cloning the remote Git repository
[2021-09-07T15:24:13.580Z] Avoid fetching tags
[Pipeline] }
[Pipeline] // stage
[Pipeline] stage
[Pipeline] { (Declarative: Post Actions)
[Pipeline] script
[Pipeline] {
[Pipeline] notifyBitbucket
[2021-09-07T15:24:13.714Z] found no commit info
[Pipeline] }
[Pipeline] // script
[Pipeline] script
[Pipeline] {
[Pipeline] deleteDir
[Pipeline] }
[Pipeline] // script
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // timestamps
[Pipeline] }
[Pipeline] // withEnv
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
Also:   hudson.remoting.Channel$CallSiteStackTrace: Remote call to channel
		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1800)
		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:356)
		at hudson.remoting.Channel.call(Channel.java:1001)
		at com.datapipe.jenkins.vault.credentials.common.VaultHelper.getVaultSecret(VaultHelper.java:42)
		at com.datapipe.jenkins.vault.credentials.common.AbstractVaultBaseStandardCredentials.getVaultSecretKeyValue(AbstractVaultBaseStandardCredentials.java:75)
		at com.datapipe.jenkins.vault.credentials.common.VaultUsernamePasswordCredentialImpl.getUsername(VaultUsernamePasswordCredentialImpl.java:61)
		at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.createUsernameFile(CliGitAPIImpl.java:2390)
		at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:2043)
		at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$500(CliGitAPIImpl.java:86)
		at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:608)
		at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:837)
		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:160)
		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:153)
		at hudson.remoting.UserRequest.perform(UserRequest.java:211)
		at hudson.remoting.UserRequest.perform(UserRequest.java:54)
		at hudson.remoting.Request$2.run(Request.java:376)
		at hudson.remoting.InterceptingExecutorService.lambda$wrap$0(InterceptingExecutorService.java:78)
Also:   hudson.remoting.Channel$CallSiteStackTrace: Remote call to secret_managment_agent_1
		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1800)
		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:356)
		at hudson.remoting.Channel.call(Channel.java:1001)
		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.execute(RemoteGitImpl.java:145)
		at sun.reflect.GeneratedMethodAccessor520.invoke(Unknown Source)
		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.lang.reflect.Method.invoke(Method.java:498)
		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.invoke(RemoteGitImpl.java:131)
		at com.sun.proxy.$Proxy116.execute(Unknown Source)
		at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1223)
		at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1301)
		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:129)
		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:97)
		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:84)
		at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
		at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
com.datapipe.jenkins.vault.exception.VaultPluginException: The credential id was not configured - please specify the credentials to use.
	at com.datapipe.jenkins.vault.credentials.common.VaultHelper.retrieveVaultCredentials(VaultHelper.java:131)
	at com.datapipe.jenkins.vault.credentials.common.VaultHelper.access$100(VaultHelper.java:29)
	at com.datapipe.jenkins.vault.credentials.common.VaultHelper$SecretRetrieve.call(VaultHelper.java:104)
	at com.datapipe.jenkins.vault.credentials.common.VaultHelper$SecretRetrieve.call(VaultHelper.java:51)
	at hudson.remoting.UserRequest.perform(UserRequest.java:211)
	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
	at hudson.remoting.Request$2.run(Request.java:376)
	at hudson.remoting.InterceptingExecutorService.lambda$wrap$0(InterceptingExecutorService.java:78)
	at org.jenkinsci.remoting.CallableDecorator.call(CallableDecorator.java:18)
	at hudson.remoting.CallableDecoratorList.lambda$applyDecorator$0(CallableDecoratorList.java:19)
	at jenkins.util.ContextResettingExecutorService$2.call(ContextResettingExecutorService.java:46)
	at jenkins.security.ImpersonatingExecutorService$2.call(ImpersonatingExecutorService.java:80)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Finished: FAILURE

### Results

Expected result:

The same behavior as withVault: if there are Folder settings, it uses them and no global setting needed. withVault works as expected.

Actual result:

"Vault Username-Password Credential" use only global Jenkins settings (/configure) and ignore Folder setting. It makes the given feature useless.

### PR DESCRIPTION
**Pull request by hchakrouun**
The modifications I introduced in this pull request are about getting the right vault configuration from the right context, so when I fill in the vault config for a multibranch pipeline then I test the retrieval of my secret from vault by the "Test Vault Secrets retrieval" button, It works while before it always using the global configuration because we were missing the credentials' context when we're looking for the vault configuration.
Now, the "Test Vault Secrets retrieval" button and the pipelines (Jenkinsfile) use the right vault config as they look for it in the right order in the chain (From pipeline => possible parents (folder Item) ==> Jenkins global config).

I also added those lines in: https://github.com/jenkinsci/hashicorp-vault-plugin/pull/195/commits/ae364e20666fdfa3f47598483865ea00c6f91c49#diff-400f26e4312f4dd4af5c888bba9019f1ef6db32fb9186de9a619b151ac115b3eR254 , so we can define a **VAULT_ADDR** env variable in the Jenkins global configuration (**Global properties** > **Environment variables** section) so that we can use it whenever the vault address is never defined, and it is available as normal environment variable in Jenkinsfiles.

I also want to mention that the **KeyStore Key** field of **Vault Certificate Credential** is not saved when saving the credential, It is the default value **keystore** that is always used, so I made some changes to the jelly template **https://github.com/jenkinsci/hashicorp-vault-plugin/pull/195/files#diff-3caead88055864dfe6bb98fdec2389f054f16220bd9148b2a3644e11ec4cf1baR14** to show it to the user as a default value so he can use it as a key of a certificate in Vault.

This pull request:
 - Closes #191 
 - Closes #124 
 - Closes #134 

- [ ] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by jetersen** [comment]
@hchakrouun can you fix the checkstyle and spotbug issues?

**Comment by hchakrouun** [comment]
> @hchakrouun can you fix the checkstyle and spotbug issues?

Yes I am working on that

**Comment by hchakrouun** [comment]
I fixed the checkSyle using eclipse plugin that uses the google-checks.xml until everything is ok in my editor, but when I pushed the mofifications, It still reports that **import com.bettercloud.vault.SslConfig;** in the imports bellow is not respecting the order, I don't know how can I fix this, is it about a version of a checkStyle file used or something I ignore ?

`package com.datapipe.jenkins.vault.configuration;

import static hudson.Util.fixEmptyAndTrim;

import com.bettercloud.vault.SslConfig;
import com.bettercloud.vault.VaultConfig;
import com.bettercloud.vault.VaultException;
import com.cloudbees.plugins.credentials.common.StandardListBoxModel;
import com.cloudbees.plugins.credentials.domains.DomainRequirement;
import com.cloudbees.plugins.credentials.domains.URIRequirementBuilder;
import com.datapipe.jenkins.vault.credentials.VaultCredential;
import com.datapipe.jenkins.vault.exception.VaultPluginException;
import edu.umd.cs.findbugs.annotations.NonNull;
import hudson.EnvVars;
import hudson.Extension;
import hudson.model.AbstractDescribableImpl;
import hudson.model.Descriptor;
import hudson.model.Item;
import hudson.security.ACL;
import hudson.slaves.EnvironmentVariablesNodeProperty;
import hudson.slaves.NodeProperty;
import hudson.slaves.NodePropertyDescriptor;
import hudson.util.DescribableList;
import hudson.util.ListBoxModel;
import hudson.util.ListBoxModel.Option;
import java.io.Serializable;
import java.util.List;
import jenkins.model.Jenkins;
import org.apache.commons.lang.StringUtils;
import org.kohsuke.accmod.Restricted;
import org.kohsuke.accmod.restrictions.NoExternalUse;
import org.kohsuke.stapler.AncestorInPath;
import org.kohsuke.stapler.DataBoundConstructor;
import org.kohsuke.stapler.DataBoundSetter;
import org.kohsuke.stapler.QueryParameter;`

**Comment by jetersen** [comment]
@hchakrouun static import goes at the bottom: 
https://github.com/jenkinsci/hashicorp-vault-plugin/blob/4b76af5de13a9541c240280dbf8b45734a286427/.mvn/checkstyle.xml#L13-L16

**Comment by hchakrouun** [comment]
The checkstyle file the plugin of my eclipse is using say that the static imports have to go first, and so did I.
But whatever, I just put it at the bottom as your checkstyle file expects and pushed

**Comment by hchakrouun** [comment]
Hi @jetersen 
As you can see, I fixed all reported failed checks.

File: src/main/java/com/datapipe/jenkins/vault/credentials/common/VaultCertificateCredentialsImpl.java [hunk 1]
```diff
@@ -132,28 +132,30 @@ private static String unwrap(@NonNull String wrapped) {
 
     @Extension
     public static class DescriptorImpl extends BaseStandardCredentialsDescriptor {
+        public static final String defaultKeyStoreKey = DEFAULT_KEYSTORE_KEY;
```
Comment (jetersen) [inline src/main/java/com/datapipe/jenkins/vault/credentials/common/VaultCertificateCredentialsImpl.java:135]: I suggest we move the DEFAULT_KEYSTORE_KEY and DEFAULT_PASSWORD_KEY to the DescriptorImpl

So they can be referred to in jelly.

and the VaultCertificateCredentialsImpl can also just refer to them from the DescriptorImpl

**Review by jetersen — CHANGES_REQUESTED** [review]
One minor issue and one ask for moving it to a separate pull request.

File: src/main/java/com/datapipe/jenkins/vault/configuration/VaultConfiguration.java [hunk 2]
```diff
@@ -246,7 +251,23 @@ public ListBoxModel doFillEngineVersionItems(@AncestorInPath Item context) {
...
+                for (NodeProperty<?> nodeProperty: globalNodeProperties) {
+                    if (nodeProperty instanceof EnvironmentVariablesNodeProperty) {
+                        EnvVars envVars = ((EnvironmentVariablesNodeProperty)nodeProperty).getEnvVars();
+                        if (envVars.containsKey("VAULT_ADDR")) {
+                            vaultConfig.address(envVars.get("VAULT_ADDR"));
+                            break;
+                        }
+                    }
+                }
+            }
```
[the remark this answers is not in the document]
Comment (hchakrouun) [reply src/main/java/com/datapipe/jenkins/vault/configuration/VaultConfiguration.java]: Yes, done

File: src/main/java/com/datapipe/jenkins/vault/credentials/common/VaultCertificateCredentialsImpl.java (hunk 1 above)
Comment (hchakrouun) [reply src/main/java/com/datapipe/jenkins/vault/credentials/common/VaultCertificateCredentialsImpl.java]: Done too

**Review by jetersen — APPROVED** [review]
LGTM

**Comment by jetersen** [comment]
link to see current commit author: https://github.com/jenkinsci/hashicorp-vault-plugin/commit/bbd8191c0aa831e4b320a1c572801670507d53c5.patch

@hchakrouun 
You might want to set up an email alias on your GitHub account 🥇
https://help.github.com/articles/adding-an-email-address-to-your-github-account/

Or look into fixing your git user.email config 😅  
https://help.github.com/articles/setting-your-commit-email-address-in-git
then you would need to amend your commits: https://stackoverflow.com/a/3042512
after changing author

to get proper credit for your commits 🏆

**Comment by hchakrouun** [comment]
Done, I changed my name and email.

**Comment by hchakrouun** [comment]
So @jetersen , are there any additional steps to validate before merging the pull request ?

**Pull request merged**